### ISSUE TITLE
调用退出登录接口后异常

### ISSUE THREAD

**Issue by su-chang** [padlocal/wechaty-puppet-padlocal#20]
```
15:59:49 VERB [CacheManager] close()
15:59:49 SILL [CacheManager] close() closing caches ...
==========logout success==========
reason : logout by self
15:59:49 SILL [CacheManager] close() cache closed.
15:59:49 SILL StateSwitch <PuppetPadlocal> off() is pending
15:59:49 VERB StateSwitch <PuppetPadlocal> off(true) <- (pending)
15:59:50 SILL Contact constructor(undefined)
15:59:50 SILL Contact ready() @ Puppet#0<PuppetPadlocal>() with id="undefined"
15:59:50 SILL StateSwitch <PuppetPadlocal> off() is true
15:59:50 WARN [PuppetPadlocal] stop() is called on a OFF puppet. await ready(off) and return.
15:59:50 VERB StateSwitch <PuppetPadlocal> ready(off, false)
15:59:50 SILL StateSwitch <PuppetPadlocal> ready(off, false) resolved.
15:59:50 VERB Contact ready() this.wechaty.puppet.contactPayload(undefined) exception: no id
15:59:50 ERR Config ###########################
15:59:50 ERR Config unhandledRejection: Error: no id
    at PuppetPadlocal.<anonymous> (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/wechaty-puppet/src/puppet.ts:684:13)
    at Generator.next (<anonymous>)
    at /Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/wechaty-puppet/dist/src/puppet.js:8:71
    at new Promise (<anonymous>)
    at __awaiter (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/wechaty-puppet/dist/src/puppet.js:4:12)
    at PuppetPadlocal.contactPayload (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/wechaty-puppet/dist/src/puppet.js:458:16)
    at WechatifiedContactSelf.ready (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/wechaty/src/user/contact.ts:829:48)
    at PuppetPadlocal.puppet.on (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/wechaty/src/wechaty.ts:474:27)
    at PuppetPadlocal.emit (events.js:198:13)
    at PuppetPadlocal.EventEmitter.emit (domain.js:448:20)
    at PadLocalClient._client.on (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/wechaty-puppet-padlocal/src/puppet-padlocal.ts:1337:12)
    at PadLocalClient.emit (events.js:198:13)
    at PadLocalClient.EventEmitter.emit (domain.js:448:20)
    at PadLocalClient.emit (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/padlocal-client-ts/src/PadLocalClient.ts:28:18)
    at Request.ret.onSystemEventCallback (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/padlocal-client-ts/src/PadLocalClient.ts:98:14)
    at Request.<anonymous> (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/padlocal-client-ts/src/Request.ts:217:35)
    at Generator.next (<anonymous>)
    at /Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/padlocal-client-ts/dist/Request.js:8:71
    at new Promise (<anonymous>)
    at __awaiter (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/padlocal-client-ts/dist/Request.js:4:12)
    at Request._onServerMessage (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/padlocal-client-ts/dist/Request.js:139:16)
    at ClientDuplexStreamImpl.Request._grpcStream.on (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/padlocal-client-ts/src/Request.ts:55:12)
    at ClientDuplexStreamImpl.emit (events.js:198:13)
    at ClientDuplexStreamImpl.EventEmitter.emit (domain.js:448:20)
    at addChunk (_stream_readable.js:288:12)
    at readableAddChunk (_stream_readable.js:269:11)
    at ClientDuplexStreamImpl.Readable.push (_stream_readable.js:224:10)
    at Object.onReceiveMessage (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/@grpc/grpc-js/src/client.ts:657:16)
    at Object.onReceiveMessage (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/@grpc/grpc-js/src/client-interceptors.ts:373:49)
    at process.nextTick (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/@grpc/grpc-js/src/call-stream.ts:313:22)
    at process._tickCallback (internal/process/next_tick.js:61:11) [object Promise]
15:59:50 ERR Config ###########################
15:59:50 ERR Config process.on(unhandledRejection) promise.catch(no id)
Config Error: no id
    at PuppetPadlocal.<anonymous> (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/wechaty-puppet/src/puppet.ts:684:13)
    at Generator.next (<anonymous>)
    at /Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/wechaty-puppet/dist/src/puppet.js:8:71
    at new Promise (<anonymous>)
    at __awaiter (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/wechaty-puppet/dist/src/puppet.js:4:12)
    at PuppetPadlocal.contactPayload (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/wechaty-puppet/dist/src/puppet.js:458:16)
    at WechatifiedContactSelf.ready (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/wechaty/src/user/contact.ts:829:48)
    at PuppetPadlocal.puppet.on (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/wechaty/src/wechaty.ts:474:27)
    at PuppetPadlocal.emit (events.js:198:13)
    at PuppetPadlocal.EventEmitter.emit (domain.js:448:20)
    at PadLocalClient._client.on (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/wechaty-puppet-padlocal/src/puppet-padlocal.ts:1337:12)
    at PadLocalClient.emit (events.js:198:13)
    at PadLocalClient.EventEmitter.emit (domain.js:448:20)
    at PadLocalClient.emit (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/padlocal-client-ts/src/PadLocalClient.ts:28:18)
    at Request.ret.onSystemEventCallback (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/padlocal-client-ts/src/PadLocalClient.ts:98:14)
    at Request.<anonymous> (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/padlocal-client-ts/src/Request.ts:217:35)
    at Generator.next (<anonymous>)
    at /Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/padlocal-client-ts/dist/Request.js:8:71
    at new Promise (<anonymous>)
    at __awaiter (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/padlocal-client-ts/dist/Request.js:4:12)
    at Request._onServerMessage (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/padlocal-client-ts/dist/Request.js:139:16)
    at ClientDuplexStreamImpl.Request._grpcStream.on (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/padlocal-client-ts/src/Request.ts:55:12)
    at ClientDuplexStreamImpl.emit (events.js:198:13)
    at ClientDuplexStreamImpl.EventEmitter.emit (domain.js:448:20)
    at addChunk (_stream_readable.js:288:12)
    at readableAddChunk (_stream_readable.js:269:11)
    at ClientDuplexStreamImpl.Readable.push (_stream_readable.js:224:10)
    at Object.onReceiveMessage (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/@grpc/grpc-js/src/client.ts:657:16)
    at Object.onReceiveMessage (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/@grpc/grpc-js/src/client-interceptors.ts:373:49)
    at process.nextTick (/Users/suchang/Desktop/PROJECT/Test/testPuppet/node_modules/@grpc/grpc-js/src/call-stream.ts:313:22)
    at process._tickCallback (internal/process/next_tick.js:61:11)
(node:19184) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 3)
```

**Comment by padlocal** [comment]
Can you provide your test case code?

```
15:59:50 WARN [PuppetPadlocal] stop() is called on a OFF puppet. await ready(off) and return.
15:59:50 VERB StateSwitch <PuppetPadlocal> ready(off, false)
15:59:50 SILL StateSwitch <PuppetPadlocal> ready(off, false) resolved.

-> 15:59:50 VERB Contact ready() this.wechaty.puppet.contactPayload(undefined) exception: no id

15:59:50 ERR Config ###########################
```

I am not sure why contactPayload will be called after logout.

**Comment by su-chang** [comment]
### Test Case
```ts
bot.logout()
```

I think maybe you miss params payload for `logout` event.

### Wechaty process logout event
```ts
case 'logout':
  puppet.on('logout', async payload => {
    const contact = this.ContactSelf.load(payload.contactId) // undefined
    await contact.ready()
    this.emit('logout', contact, payload.data)
  })
  break
```

**Comment by padlocal** [comment]
Still can not reproduce this issue. logout is handled as following in PadLocal Puppet:

```
public async logout(): Promise<void> {
    if (!this.id) {
      throw new Error("logout before login?");
    }

    await this._client!.api.logout();

    this.emit("logout", { contactId: this.id, data: "logout by self" });

    await this.stop();
}
```

```
this._client.on("kickout", async (_detail: KickOutEvent) => {
    this.emit("logout", { contactId: this.id!, data: JSON.stringify(_detail) });

    await this.stop();
});
```

I think the parameters are well supplied.

**Comment by su-chang** [comment]
Do you listen on `logout` event in `wechaty` ?

```ts
bot.on('logout', (user, reason) => {
  console.log(`==========logout success==========`);
  if (reason) {
    console.log(`reason : ${reason}`);
  }
})
```

**Comment by padlocal** [comment]
fixed in wechaty-puppet-padlocal@0.2.11
